### PR TITLE
fix: Fix race-condition in reindexing

### PR DESCRIPTION
**Pull request by mykola-vrmchk**

### CODE REVIEW

File: packages/core/echo/indexing/src/index-metadata-store.ts [hunk 1]
```diff
@@ -75,8 +98,20 @@ export class IndexMetadataStore {
     this.dirty.emit();
   }
 
-  async markClean(id: string, lastIndexedHash: string) {
+  async markClean(id: ObjectPointerEncoded, lastIndexedHash: ConcatenatedHeadHashes) {
     await this._lastIndexed.put<string, string>(id, lastIndexedHash, { valueEncoding: 'json' });
     this.clean.emit();
   }
+
+  // markClean(idToLastHash: Map<ObjectPointerEncoded, ConcatenatedHeadHashes>, batch: BatchLevel) {
```
Comment (dmaretskyi) [inline packages/core/echo/indexing/src/index-metadata-store.ts:106]: remove or, better, integrate the batched version everywhere

File: packages/core/echo/indexing/src/index-metadata-store.ts [hunk 2]
```diff
@@ -75,8 +98,20 @@ export class IndexMetadataStore {
...
+  //   for (const [id, lastIndexedHash] of idToLastHash.entries()) {
+  //     batch.put<string, string>(id, lastIndexedHash, { valueEncoding: 'json', sublevel: this._lastIndexed });
+  //   }
+  // }
+
+  dropFromClean(ids: ObjectPointerEncoded[], batch: BatchLevel) {
```
Comment (dmaretskyi) [inline packages/core/echo/indexing/src/index-metadata-store.ts:112]: add jsdoc to explain when this is called

File: packages/sdk/client/src/tests/indexing.test.ts [hunk 3]
```diff
@@ -17,30 +17,70 @@ import { log } from '@dxos/log';
...
+  const initClient = async (services: ClientServicesProvider) => {
+    const client = new Client({ services });
+    await client.initialize();
+    if (!client._graph.runtimeSchemaRegistry.isSchemaRegistered(ContactType)) {
+      client._graph.runtimeSchemaRegistry.registerSchema(ContactType);
+    }
+    return client;
+  };
+
+  const addJohn = async (space: Space) => {
```
Comment (dmaretskyi) [inline packages/sdk/client/src/tests/indexing.test.ts:39]: addContact(space, name)

File: packages/sdk/client/src/tests/indexing.test.ts [hunk 4]
```diff
@@ -17,30 +17,70 @@ import { log } from '@dxos/log';
...
+    await space.waitUntilReady();
+    const contact = create(ContactType, { name: john, identifiers: [] });
+    space.db.add(contact);
+    await space.db.flush();
+    return contact;
+  };
+
+  const queryIndexedJohn = async (space: Space) => {
```
Comment (dmaretskyi) [inline packages/sdk/client/src/tests/indexing.test.ts:47]: queryIndexedContact(space, name)

**Comment by mykola-vrmchk** [comment]
* **#6419** <a href="https://app.graphite.dev/github/pr/dxos/dxos/6419?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mykola-vrmchk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Pull request merged**